packages/frontend/tests/unit/service/channels/channel-verifier-test.ts

```diff
@@ -0,0 +1,755 @@
+import { sign, hash } from 'emberclear/workers/crypto/utils/nacl';
+
+module('Unit | Service | channels/channel-verifier', function (hooks) {
+  setupTest(hooks);
+  clearLocalStorage(hooks);
+
+  let service!: ChannelVerifier;
+
+  hooks.beforeEach(function () {
+    service = getService('channels/channel-verifier');
+  });
+
+  module('starting channel', function () {
+    module('invalid', function () {
+      test('no admin', async function (assert) {
+        const store = getStore();
+        let channelContextChain = store.createRecord('channel-context-chain', {
+          members: [],
+        });
+        assert.notOk(await service.verify(channelContextChain));
+      });
+
+      test('no members', async function (assert) {
+        const store = getStore();
+        let admin = await buildUser('admin');
+        let channelContextChain = store.createRecord('channel-context-chain', {
+          admin: admin,
+          members: [],
+        });
+        assert.notOk(await service.verify(channelContextChain));
+      });
+
+      test('more than 1 member', async function (assert) {
+        const store = getStore();
+        let admin = await buildUser('admin');
+        let member1 = await buildUser('member1');
+        let channelContextChain = store.createRecord('channel-context-chain', {
+          admin: admin,
+          members: [admin, member1],
+        });
+        assert.notOk(await service.verify(channelContextChain));
+      });
+
+      test('member is not admin', async function (assert) {
+        const store = getStore();
+        let admin = await buildUser('admin');
+        let member1 = await buildUser('member');
+        let channelContextChain = store.createRecord('channel-context-chain', {
+          admin: admin,
+          members: [member1],
+        });
+        assert.notOk(await service.verify(channelContextChain));
+      });
+    });
+
+    test('valid', async function (assert) {
+      const store = getStore();
+      let admin = await buildUser('admin');
+      let channelContextChain = store.createRecord('channel-context-chain', {
+        admin: admin,
+        members: [admin],
+      });
+      assert.ok(await service.verify(channelContextChain));
+    });
+  });
+
+  module('subsequent chains', function () {
+    module('valid', function () {
+      test('add a member', async function (assert) {
+        const store = getStore();
+        let admin = await buildUser('admin');
+        let member1 = await buildUser('member1');
+        let vote = store.createRecord('vote-chain', {
+          yes: [admin],
+          no: [],
+          remaining: [],
+          action: VOTE_ACTION.ADD,
+          target: member1,
+          key: admin,
+          previousVoteChain: undefined,
+          signature: undefined,
+        });
+        vote.signature = await sign(await hash(generateSortedVote(vote)), admin.privateSigningKey);
```

Comment:
should this be extracted to a helper that just takes vote and user?

```ts
let signature = await signatureOf(vote, admin)
```

Reply:
addressed in https://github.com/NullVoxPopuli/emberclear/pull/738/commits/d4cba3fd1da38e46cb82af774a584a0f014ed56d